webapp/src/test/java/com/google/sps/servlets/RunValidatorTest.java

```diff
@@ -38,4 +51,107 @@ public void testBartGtfsFeed() {
     assertThat(json.exportJson()).contains("\"code\":\"unknown_column\",\"totalNotices\":5");
     assertThat(json.exportJson()).contains("\"code\":\"unexpected_file\",\"totalNotices\":5");
   }
+
+  @Test
+  public void testValidFileUpload() throws IOException, ServletException {
+    MockHttpServletRequest request = new MockHttpServletRequest();
+    MockHttpServletResponse response = new MockHttpServletResponse();
+
+    // Load file from resources
+    String resourceName = "/SAMPLE.zip";
+    String partName = "SAMPLE";
+    byte[] fileContent =
+        FileCopyUtils.copyToByteArray(getClass().getResourceAsStream(resourceName));
+    // Create part & entity from resource
+    Part[] parts =
+        new Part[] {new FilePart(partName, new ByteArrayPartSource(resourceName, fileContent))};
+    MultipartRequestEntity multipartRequestEntity =
+        new MultipartRequestEntity(parts, new PostMethod().getParams());
+    // Convert to bytes and write to entity
+    ByteArrayOutputStream requestContent = new ByteArrayOutputStream();
+    multipartRequestEntity.writeRequest(requestContent);
+    // Update request
+    request.setContent(requestContent.toByteArray());
+    request.setContentType(multipartRequestEntity.getContentType());
+
+    new UploadServlet().doPost(request, response);
+
+    String validUploadMsg = "Upload has been done successfully!";
+    assertThat(response.getContentAsString()).contains(validUploadMsg);
+    assertThat(response.getContentAsString()).contains("invalid_row_length");
+  }
+
+  @Test
+  public void testInvalidFileUpload() throws IOException, ServletException {
```

Comment:
What is the difference for upload functionality if the zip file has missing portions?

Reply:
Do you mean if the zip only has some required files but not all ? The validator will return the notice: 
"code":"missing_required_file","totalNotices":5. It appears there are 5 required .txt files for the validation to work

Reply:
I am not sure this test is needed.

We are testing upload functionality (not validator).

Both this case and the case above present test cases where the file was uploaded correctly. I see no additional value in this test.

Reply:
This case tests when the user uploads an incorrect file (html) instead of zip. This was to check that this would be dealt with

Reply:
Ok true. I will move this to the runValidator branch

Reply:
make sense. At that point it needs to run only through the validator, not including any upload.